dspace-spring-rest/src/test/java/org/dspace/statistics/MockSolrLoggerServiceImpl.java

```diff
@@ -33,20 +46,67 @@ public MockSolrLoggerServiceImpl() {
 
     @Override
     public void afterPropertiesSet() throws Exception {
+        // Initialize our service with a Mock Solr statistics core
         mockSolrServer = new MockSolrServer("statistics");
         solr = mockSolrServer.getSolrServer();
+    }
+
+    @Override
+    public void postView(DSpaceObject dspaceObject, HttpServletRequest request,
+                         EPerson currentUser) {
+        // Load our FakeDatabaseReader before each view event is logged
+        loadFakeDatabaseReader();
+        // Call overridden method
+        super.postView(dspaceObject, request, currentUser);
+    }
+
+    @Override
+    public void postView(DSpaceObject dspaceObject,
+                         String ip, String userAgent, String xforwardedfor, EPerson currentUser) {
+        // Load our FakeDatabaseReader before each view event is logged
+        loadFakeDatabaseReader();
+        // Call overridden method
+        super.postView(dspaceObject, ip, userAgent, xforwardedfor, currentUser);
+    }
 
-        new FakeDatabaseReader(); // Activate fake
-        new FakeDatabaseReader.Builder(); // Activate fake
-        String locationDbPath = configurationService.getProperty("usage-statistics.dbfile");
-        File locationDb = new File(locationDbPath);
-        locationDb.createNewFile();
-        locationService = new DatabaseReader.Builder(locationDb).build();
+    @Override
+    public void postSearch(DSpaceObject resultObject, HttpServletRequest request, EPerson currentUser,
+                           List<String> queries, int rpp, String sortBy, String order, int page, DSpaceObject scope) {
+        // Load our FakeDatabaseReader before each search event is logged
+        loadFakeDatabaseReader();
+        // Call overridden method
+        super.postSearch(resultObject, request, currentUser, queries, rpp, sortBy, order, page, scope);
+    }
+
+    @Override
+    public void postWorkflow(UsageWorkflowEvent usageWorkflowEvent) throws SQLException {
+        // Load our FakeDatabaseReader before each workflow event is logged
+        loadFakeDatabaseReader();
+        // Call overridden method
+        super.postWorkflow(usageWorkflowEvent);
     }
 
     @Override
     public void destroy() throws Exception {
         mockSolrServer.destroy();
     }
 
+    /**
+     * Load / activate our FakeDatabaseReader which uses JMockit to replace specific
+     * methods of the DatabaseReader. This MUST be called for each method that uses
+     * DatabaseReader as JMockit fake classes are only in effect for tests in which
+     * they are defined: http://jmockit.github.io/tutorial/Faking.html
+     */
+    private void loadFakeDatabaseReader() {
+        try {
```

Comment:
Maybe we should surround this  `try` block with an `if (locationService == null)` statement so that we only do this initialization once.

Reply:
Hi @tomdesair : I tried adding that extra `if` today, and unfortunately it doesn't work (the first test using Stats succeeds, but the second one fails). 

From what I can tell, the JMockit Fake Class (`FakeDatabaseReader`) seems to only "live" for one method call.  So, it works for the first test method, but the second test method tries to use the *real* DatabaseReader, which causes test failures. 

It's possible there's a better way to implement this code in general, but I've already sunk enough time into this -- and the result is already a massive speed improvement. So, I suggest we just consider this code "good enough".

Reply:
Ok thanks for trying